wai-language-server/src/main.rs

```diff
@@ -49,16 +110,296 @@ impl LanguageServer for Backend {
+    fn from(val: BindingLanguage) -> Self {
+        match val {
+            BindingLanguage::Rust => "Rust".to_string(),
+            BindingLanguage::JavaScript => "Javascript".to_string(),
+            BindingLanguage::Python => "Python".to_string(),
+            BindingLanguage::C => "C".to_string(),
+        }
+    }
+}
+
+impl From<&str> for BindingLanguage {
+    fn from(s: &str) -> Self {
+        match s {
+            "Rust" => BindingLanguage::Rust,
+            "Javascript" => BindingLanguage::JavaScript,
+            "Python" => BindingLanguage::Python,
+            "C" => BindingLanguage::C,
+            _ => panic!("Unknown binding language"),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+enum GenerationDirection {
+    Guest,
+    Host,
+}
+
+impl From<GenerationDirection> for String {
+    fn from(val: GenerationDirection) -> Self {
+        match val {
+            GenerationDirection::Guest => "Guest".to_string(),
+            GenerationDirection::Host => "Host".to_string(),
+        }
+    }
+}
+
+impl From<&str> for GenerationDirection {
+    fn from(s: &str) -> Self {
+        match s {
+            "Guest" => GenerationDirection::Guest,
+            "Host" => GenerationDirection::Host,
+            _ => panic!("Unknown generation type"),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+enum GenerationType {
+    Import,
+    Export,
+}
+
+impl From<GenerationType> for String {
+    fn from(val: GenerationType) -> Self {
+        match val {
+            GenerationType::Import => "Import".to_string(),
+            GenerationType::Export => "Export".to_string(),
+        }
+    }
+}
+
+impl From<&str> for GenerationType {
+    fn from(s: &str) -> Self {
+        match s {
+            "Import" => GenerationType::Import,
+            "Export" => GenerationType::Export,
+            _ => panic!("Unknown generation type"),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct WAIGeneratorParams {
+    generation_direction: GenerationDirection,
+    generation_type: GenerationType,
+    binding_language: BindingLanguage,
+    file_name: String,
+    file_content: String,
+}
+
+impl Backend {
+    async fn generate_wai_code(&self, params: WAIGeneratorParams) -> Result<()> {
+        let WAIGeneratorParams {
+            binding_language,
+            generation_direction,
+            generation_type,
+            file_name,
+            file_content,
+        } = params;
+
+        let mut files = Files::default();
+
+        let interface = Interface::parse(&file_name, &file_content).unwrap();
+
+        let (imports, exports): (Vec<Interface>, Vec<Interface>) = match generation_type {
+            GenerationType::Import => (vec![interface], vec![]),
+            GenerationType::Export => (vec![], vec![interface]),
+        };
+
+        match generation_direction {
+            GenerationDirection::Guest => match binding_language {
```

Comment:
The nested `match` statements are kinda verbose and error-prone (e.g. we need to handle every single combination and "somehow" tell the user when they've used an invalid combination).

A better solution could be something like this:

```rs
enum Direction {
    Guest,
    Host,
}

enum Language {
    Rust,
    JavaScript,
    Python,
    C,
}

enum Type {
    Import,
    Export,
}

type GeneratorFunc = fn(&str, &mut Files, Type);

struct Generator {
    language: Language,
    direction: Direction,
    generate: GeneratorFunc,
}

// Solution A

const GENERATORS: &[Generator] = &[Generator {
    language: Language::Rust,
    direction: Host,
    generate: |src: &str, files: &mut Files, ty: Type| {
        let opts = RustGuestGenOpts {
            rustfmt: true,
            force_generate_structs: true,
            ..Default::default()
        };
        let mut gen = opts.build();
        gen.generate_all(&imports, &exports, &mut files)
    },
}];

fn get_generator(language: Language, direction: Direction) -> Option<&'static Generator> {
  GENERATORS.iter().find(|g| g.language == language && g.direction == direction)
}

// Solution B

fn main() {
    let generators: HashMap<(Language, Direction), GeneratorFunc> = HashMap::new();

    generators.insert(
        (Language::Rust, Direction::Host),
        |src: &str, files: &mut Files, ty: Type| {
            let opts = RustGuestGenOpts {
                rustfmt: true,
                force_generate_structs: true,
                ..Default::default()
            };
            let mut gen = opts.build();
            gen.generate_all(&imports, &exports, &mut files)
        },
    );
}

```